service/src/test/java/bio/terra/externalcreds/services/ProviderServiceTest.java

```diff
@@ -123,6 +129,139 @@ void testDeleteLinkLinkNotFound() {
+            .thenReturn(
+                OAuth2AccessTokenResponse.withToken("token").tokenType(TokenType.BEARER).build());
+        when(fenceAccountKeyServiceMock.getFenceAccountKey(linkedAccount))
+            .thenReturn(Optional.of(key));
+
+        //  Mock the server response
+        mockServer
+            .when(
+                HttpRequest.request(revocationPath)
+                    .withMethod("POST")
+                    .withQueryStringParameters(expectedParameters))
+            .respond(HttpResponse.response().withStatusCode(HttpStatus.OK.value()));
+
+        mockServer
+            .when(HttpRequest.request(keyRevocationPath).withMethod("DELETE"))
+            .respond(HttpResponse.response().withStatusCode(HttpStatus.OK.value()));
+
+        providerService.deleteLink(linkedAccount.getUserId(), linkedAccount.getProvider());
+        verify(linkedAccountServiceMock)
+            .deleteLinkedAccount(linkedAccount.getUserId(), linkedAccount.getProvider());
+        mockServer.verify(
+            HttpRequest.request(keyRevocationPath + "/" + privateKeyId).withMethod("DELETE"),
+            VerificationTimes.exactly(1));
+      }
+    }
+
+    @Test
+    void testDeleteFenceLinkNoKeyEndpoint() throws IOException {
+      try (var mockServer = ClientAndServer.startClientAndServer()) {
+        var revocationPath = "/test/revoke/";
+        var linkedAccount =
+            TestUtils.createRandomLinkedAccount().withId(1).withProvider(Provider.FENCE);
+
+        var key =
+            TestUtils.createRandomFenceAccountKey()
+                .withLinkedAccountId(linkedAccount.getId().get());
+        var privateKeyId = objectMapper.readTree(key.getKeyJson()).get("private_key_id").asText();
+        var keyRevocationPath = "/test/key";
+
+        var providerInfo =
+            TestUtils.createRandomProvider()
+                .setRevokeEndpoint(
+                    "http://localhost:" + mockServer.getPort() + revocationPath + "?token=%s");
+
+        var expectedParameters =
+            List.of(
+                new Parameter("token", linkedAccount.getRefreshToken()),
+                new Parameter("client_id", providerInfo.getClientId()),
+                new Parameter("client_secret", providerInfo.getClientSecret()));
+
+        when(externalCredsConfigMock.getProviderProperties(linkedAccount.getProvider()))
+            .thenReturn(providerInfo);
+
+        when(linkedAccountServiceMock.getLinkedAccount(
+                linkedAccount.getUserId(), linkedAccount.getProvider()))
+            .thenReturn(Optional.of(linkedAccount));
+        when(linkedAccountServiceMock.deleteLinkedAccount(
+                linkedAccount.getUserId(), linkedAccount.getProvider()))
+            .thenReturn(true);
+
+        when(providerOAuthClientCacheMock.getProviderClient(linkedAccount.getProvider()))
+            .thenReturn(createClientRegistration(linkedAccount.getProvider()));
+
+        when(oAuth2ServiceMock.authorizeWithRefreshToken(
```

Comment:
It looks like these tests mock some of the same things. I think it would be easier to read if those were split out into helper functions, but I don't consider it to be a blocker!

Reply:
Yeah, lots of repeated code. I cleaned it up!